feat(search) Add functions for creating conditions on nullable arrays

markstory opened the pull request:
The current IN conditions we use on `exception_stacks_mechanism.handled` are not what we want in the product. This field has an unconventional data model where undefined/None mean the same thing as True, and not False. This means that end user queries like `error.handled:0` and `error.handled:1` do not work as required. These new functions make it possible for sentry to generate the correct conditions.

I was not able to use the `CustomFunction` abstraction for this because it cannot currently parse the required lambda inside arrayExists().

Refs https://www.notion.so/sentry/error-handled-conditions-03bea1a3ad3d45639fc3976465638ccf

markstory committed:
feat(search) Add functions for creating conditions on nullable arrays

The current IN conditions we use on `exception_stacks_mechanism.handled`
are not what we want in the product. This field has an unconventional
data model where undefined/None mean the same thing as True, and not
False. This means that end user queries like `error.handled:0` and
`error.handled:1` do not work as required. These new functions make it
possible for sentry to generate the correct conditions.

I was not able to use the `CustomFunction` abstraction for this because
it cannot currently parse the required lambda inside arrayExists().

Refs https://www.notion.so/sentry/error-handled-conditions-03bea1a3ad3d45639fc3976465638ccf

evanh reviewed (commented):
Makes sense to me. Too bad about CustomFunction though.

markstory committed:
Merge branch 'master' into feat/array-exists

fpacifici commented:
I have a couple of questions on what we are trying to accomplish that I cannot get from the notion document.
This is the code that fills in the array:
https://github.com/getsentry/snuba/blob/master/snuba/datasets/events_processor_base.py#L295

For each frame, if the input is `1`, `yes` or `true` what is stored is 1 same for the value 0. Everything else (NULL and invalid input) is NULL and it means we do not have data for the `handled` field. It does not mean true nor false. 

I don't think we should conflate the NULL value with the True value. It does not mean the exception was handled. It means we do not know. 

- when you say that `undefined/None mean the same thing as True, and not False` are you talking about clickhouse or are you talking about the product expectation? The first would not be accurate.
- what are the conditions you want to use in the product? I assume "is any exception handled", "is any exception not handled" are part of them. Are there others ?

markstory commented:
> when you say that undefined/None mean the same thing as True, and not False are you talking about clickhouse or are you talking about the product expectation? The first would not be accurate.

This is the product expectation that the mobile folks have. The intended behavior they want is outlined in https://www.notion.so/sentry/Crashes-in-SDKs-and-Product-fa249e31d6274b66b1bfff966e00b668#95e2fa1035fe4a76a77208cf83a1b8ef

> what are the conditions you want to use in the product? I assume "is any exception handled", "is any exception not handled" are part of them. Are there others ?

The only use case we have for these functions presently is for building conditions on the handled field.

fpacifici commented:
> The only use case we have for these functions presently is for building conditions on the handled field.

Ok, follow up question. What are the states you want the `exception_stacks_mechanism.handled` in the abstract snuba data model to have? It seems to me that this field would be exposed as an individual boolean field of an event while it actually is an array of ternary fields on the event.
So it is important to decide how we model this in the logical data model since that decision impacts all products using it:
- do we want to deal with it at event level, as a field existing in every stack (so with the same cardinality of stacks) or both (some features see the array but we have some special cases where queries that want to see an aggregated view) ?
- if it is always the first then we should change the schema, but does that apply to the other stacks fields as well?
- if it is the third, the approach is fine (I still do not understand the business logic that makes a a null count as handled but that is a product decision), though this logic seems very specific to some assumptions we make around this specific field, I would not consider it a general function which is hard to grasp outside of this context. I would give the function a name that is related with the handled field like `isHandled`. We may generalize them later if it turns out to be useful. Making it more general will be easier than making it more specific.

mitsuhiko commented:
These are all good questions largely because they also affect the product. So with the power of hindsight I think handled is a very odd field at the moment because it's attached to the exception but from the product perspective it has not a lot of value here. In fact the product most likely would draw a conclusion if a specific exception is unhandled.

On the protocol due to nested/chained exceptions the handled flag is set on the exception at the time the event is created. This means you can have multiple exceptions in a chain but really only the last thrown one is going to have the "handled" flag in the mechanism ever set to `False`.

In a way what the product is more interested in, is a specific combination of this: from the product's perspective an error event is "unhandled" if the last exception's mechanism is not handled.

I am not sure what other fields we're currently indexing that are similar in nature. There are for sure a few fields which could have the same situation. For instance the C `errno` is also stored on the mechanism but there might be multiple exceptions in a chain that might have that attached. It's a more pressing issue with .NET where chained exceptions are very common and the useful signal is on the first and not the last exception in a chain.

Aside: from a user's perspective in the product the "negation" is typically what they want. Users expect to search for "unhandled" errors, not "handled" events. So the fact that they have to search for "error.handled:0" is more confusing thatn to just search for "error.unhandled:yes".

markstory committed:
Rename functions

Use names that are not generalized as the behavior they provide is very
specific to the unique data model of the handled field.

markstory committed:
Merge branch 'master' into feat/array-exists

fpacifici commented:
So, this seems to point in the direction that, from the snuba logical data model stand point the field should be associated with the errors but that would create an inconsistency with the input: the field means one thing in input and a different thing in output.

So I would keep it in the array where it is in the query and I would add field specific functions to query that field in a convenient way for the user. So the data model is still intact, it is coherent between input and output and still you would have some helper functions to query what you need in a simple way. 
If we agree on this, I would suggest to rename the functions @markstory added to make them specific to the handled fields (as I suggested above). This behavior does not seem very common or generalizable. Should it become we will rename.

markstory committed:
Remove pointless parameters

These parameters don't really need to exist as they only function
correctly with a single value. I've left the column as a parameter in
case we need to use these functions on other fields and to be compatible
with snuba's query syntax.

markstory committed:
Merge branch 'master' into feat/array-exists

markstory commented:
@fpacifici I've renamed the functions and removed the 0/1 parameters as they were effectively fixed inputs for the functions to work properly. I left the column name as a parameter to make using the functions match other functions we have.

fpacifici commented on snuba/query/processors/handled_functions.py:
```diff
@@ -0,0 +1,95 @@
+                                    ),
+                                ),
+                            ),
+                            exp.parameters[0],
```
Why do you need to get the column as parameter?
I think this function should only work on a specific column per dataset. Why not providing the column name in the constructor and not allowing it to be parametric? This way there will be fewer ways to build an invalid query.

markstory replied:
Good idea. Done :+1:

fpacifici commented on snuba/query/processors/handled_functions.py:
```diff
@@ -0,0 +1,95 @@
+        def process_functions(exp: Expression) -> Expression:
+            if isinstance(exp, FunctionCall):
+                if exp.function_name == "isHandled":
+                    return FunctionCall(
```
Could you please validate the function call? For example check it has the right number of parameters ? If it is not valid please raise any `snuba.query.exceptions.InvalidExpressionException`. That will trigger an HTTP400 instead of 500.

markstory replied:
Done, I used the same signature validator that was used in CustomFunction.

fpacifici reviewed (changes requested):
Ok, almost there. A couple of comments inline

markstory committed:
Merge branch 'master' into feat/array-exists

markstory committed:
Add parameter validation and remove column parameter.

fpacifici commented on snuba/query/processors/handled_functions.py:
```diff
@@ -0,0 +1,108 @@
+            if isinstance(exp, FunctionCall):
+                if exp.function_name == "isHandled":
+                    validator = SignatureValidator([])
+                    validator.validate(exp.parameters, self.__columnset)
```
Unfortunately this raises the wrong type of exception.
InvalidFunctionCall does not extend InvalidQueryException. Thus no HTTP400.
This is how the validation logic raises the appropriate exception.
https://github.com/getsentry/snuba/blob/master/snuba/query/parser/validation/functions.py#L51-L57

markstory committed:
Raise correct exception type on validation errors.

markstory committed:
Merge branch 'master' into feat/array-exists

markstory committed:
Merge branch 'master' into feat/array-exists

markstory commented:
@fpacifici Anything left to do on this?

fpacifici reviewed (approved):
Sorry for the delay. I was out of (whatever I call) office.

markstory committed:
Merge branch 'master' into feat/array-exists
